engines/router/missionctl/log/resultlog/proto/turing/TuringResultLog.proto

```diff
@@ -5,46 +5,42 @@ package turing;
 option java_outer_classname = "TuringResultLogProto";
 option go_package = "github.com/gojek/turing";
 
-import "google/protobuf/struct.proto";
 import "google/protobuf/timestamp.proto";
 
 message Request {
-    // The header from the incoming request to the Turing router
-    map<string, google.protobuf.ListValue> header = 1;
+    // The header from the incoming request to the Turing router, in JSON format
+    string header = 1;
```

Comment:
I think map<string,string> is more commonly used data type to represent HTTP headers.

Example from Google APIs:
https://github.com/googleapis/googleapis/blob/master/google/cloud/scheduler/v1/target.proto#L61
https://github.com/googleapis/googleapis/blob/master/google/actions/sdk/v2/webhook.proto#L56

Also since according to HTTP 1.1 spec https://tools.ietf.org/html/rfc2616#page-31
http message header consists of header field name and field value, so map<string,string> looks reasonable to represend this data. I think we should use JSON string representation when the format of the data can be arbitraty object.

Reply:
Hi David, I agree that `map<string,string>` is better for this. With that, the resulting header would be of the following BQ schema: 
```
{
    "name": "header",
    "type": "RECORD",
    "mode": "REPEATED",
    "description": "The header from the incoming request to the Turing router",
    "fields": [
     {
      "name": "key",
      "type": "STRING",
      "mode": "NULLABLE"
     },
     {
      "name": "value",
      "type": "STRING",
      "mode": "NULLABLE"
     }
    ]
   }
```

And this is the excerpt from the proto's Go library:
```
Header map[string]string `protobuf:"bytes,1,rep,name=header,proto3" json:"header,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
```

They both match. But, the Go BQ library's dynamic schema inference on the `TuringResultLogMessage` struct fails ([ref](https://godoc.org/cloud.google.com/go/bigquery#StructSaver)), resulting in the below error:
```
repeated schema field header requires slice or array, but value has type map[string]string
```

I'm guessing it doesn't have extensive support for the protobuf annotations.

Our current implementation of the BQ logger uses `string` for representing the header, so I decided to keep it (not saying that this is correct or better). I don't see any major benefit of changing this to a map. However, if we feel strongly that this is the convention, I can handle it for BQ/Fluentd Loggers. Wdyt?

Reply:
The special handling would go [here](https://github.com/krithika369/turing/blob/update_turing_reslog_proto/engines/router/missionctl/log/resultlog/bigquery.go#L31), in the BQ's [ValueSaver](https://godoc.org/cloud.google.com/go/bigquery#ValueSaver) implementation. (I'm already using the ValueSaver method instead of StructSaver because the latter generates column names like `turingReqId` instead of `turing_req_id`.)

Reply:
Personally I think BigQuery specific issue should not affect the protobuf schema of our message. After all, BigQuery is only one of the many storage destinations. Some other storage destinations may have other requirements, so it's almost impossible to satisfy all the requirements unless we make all the types generic (like bytes or string). I think this responsibility should fall on the mapper/tooling to convert between Proto schema and BigQuery schema.

The benefit of having more specific type is that it is clearer from the consumer of the message the data format to expect. If we just use (json) string, without looking at the examples or documentation, these formats will be acceptable as headers
```json
{
  "Host": "http://example.com",
  "Content-Type": "json"
}
```
or 
```json
[
  {"key": "Host", "value": "http://example.com"},
  {"key": "Content-Type", "value": "json"}
]
```

But I guess if BigQuery schema compatibility (with certain tooling) is important in our case, we can use string type, but we need to provide more documentation regarding the JSON string representation of the HTTP headers.

Reply:
Ok, makes sense. I'll update the schema to `map<string, string>`.

Reply:
oh just wondering will `map<string, string>` cause issue with the schema generated from Beast?

Reply:
No, Beast supports this as expected (generated BQ schema will be `repeated` `record` with fields 'key' and 'value').